### ISSUE TITLE
Get rid of page-count based commons chunk

### ISSUE THREAD

**Issue by herrstucki** [zeit/next.js#4846]
# Feature request

## Is your feature request related to a problem? Please describe.

Currently, the logic to move things into the commons chunk is based on the count of pages (ie. files in the `pages/` directory). I think this approach doesn't scale very well and is too simplistic.

- If we have a larger app with different sets of pages that have shared dependencies, they won't be moved to up to shared chunks. For example, when 30% of pages depend on module "a".
- Even if we "cheat" by importing things into more pages, they will only get moved to the single commons chunk where they're unnecessary by all other pages (in our example, "a" will end up in commons but is not needed on 70% of pages)
- If I interpret the generated bundles correctly, there are no shared chunks created for anything that isn't a page.

## Describe the solution you'd like

- Ideally, the 30% of pages which depend on module "a" should have a shared chunk containing "a". This chunk would not be loaded/used by the other 70% of pages
- Non-page modules should also be able to have shared chunks

In short: chunk splitting should be more granular. I believe with the move to webpack 4 in https://github.com/zeit/next.js/pull/4639 this shouldn't be too hard, as it would generally mean giving more control over the chunking back to webpack.

I haven't analyzed the situation with the current canary branch, so my observations are still based on the current stable webpack 3 based approach. However, the same logic is in place at the moment:

https://github.com/zeit/next.js/blob/f4988e7fa3152920125498778dbcb9656be24072/build/webpack.js#L86-L94

If this sounds interesting @timneutkens, I'd be happy to spend some time on this by trying out a solution.

## Additional context

https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366

**Comment by vjpr** [comment]
FYI Optimization config for canary is here: https://github.com/zeit/next.js/blob/75476a91362d6dc4e9fe7bda103adeab1e81f2d8/build/webpack.js#L59

Webpack config is here: https://github.com/zeit/next.js/blob/75476a91362d6dc4e9fe7bda103adeab1e81f2d8/build/webpack.js#L106

**Comment by timneutkens** [comment]
We can enable the default behavior next to the commons behavior, I didn't on purpose as it led to a *ton* of duplicated modules on zeit.co, literally almost everything was duplicated except 3 or 4 default bundles that were generated. So it needs more investigating.

Basically, the page size went from 1MB to 12MB with prefetching 😅

**Comment by vjpr** [comment]
@timneutkens 

> Ideally, the 30% of pages which depend on module "a" should have a shared chunk containing "a". This chunk would not be loaded/used by the other 70% of pages

Is this possible? Its like a dynamic `cacheGroup` I think.

**Comment by herrstucki** [comment]
OK, sounds good, I can give it a shot over the next few days.

But ending up with _bigger_ page bundles when removing the 50%-of-pages rule sounds counter-intuitive to me 🤔 I had an app where exactly 50% of pages shared one set of dependencies and the other 50% a different one, the result being that nothing ended up in the commons chunk (but as I said, this was with the current webpack 3 based setup).

**Comment by vjpr** [comment]
Using default webpack4 config bumped my client build time from 30s to 3mins+.

**Comment by timneutkens** [comment]
Solved in https://github.com/zeit/next.js/pull/7696

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
